Source/JavaScriptCore/wasm/WasmBBQJIT.cpp

```diff
@@ -6541,41 +6568,54 @@ class BBQJIT {
         if (!condition.isNone() && !condition.isConst())
             emitMove(condition, conditionLocation);
         consume(condition);
-        currentControlData().endBlock(*this, target, results, false, condition.isNone());
 
         if (condition.isNone())
             LOG_INSTRUCTION("Branch");
         else
             LOG_INSTRUCTION("Branch", condition, conditionLocation);
 
-        if (condition.isConst() || condition.isNone())
+        if (condition.isConst() || condition.isNone()) {
+            currentControlData().flushAndSingleExit(*this, target, m_parser->expressionStack(), false, condition.isNone());
             target.addBranch(m_jit.jump()); // We know condition is true, since if it was false we would have returned early.
-        else
-            target.addBranch(m_jit.branchTest32(ResultCondition::NonZero, m_scratchGPR));
+        } else {
+            currentControlData().flushAtBlockBoundary(*this, 0, m_parser->expressionStack(), condition.isNone());
+            Jump ifNotTaken = m_jit.branchTest32(ResultCondition::Zero, m_scratchGPR);
+            currentControlData().addExit(*this, target.targetLocations(), m_parser->expressionStack());
+            target.addBranch(m_jit.jump());
+            ifNotTaken.link(&m_jit);
+            currentControlData().finalizeBlock(*this, target.targetLocations().size(), m_parser->expressionStack(), true);
+        }
+
         return { };
     }
 
     PartialResult WARN_UNUSED_RETURN addSwitch(Value condition, const Vector<ControlData*>& targets, ControlData& defaultTarget, Stack& results)
     {
+        UNUSED_PARAM(results);
         ASSERT(condition.type() == TypeKind::I32);
 
         LOG_INSTRUCTION("BrTable", condition);
 
         if (!condition.isConst())
             emitMove(condition, Location::fromGPR(m_scratchGPR));
         consume(condition);
-        currentControlData().endBlock(*this, defaultTarget, results, false, true);
 
         if (condition.isConst()) {
             // If we know the condition statically, we emit one direct branch to the known target.
             int targetIndex = condition.asI32();
-            if (targetIndex >= 0 && targetIndex < static_cast<int>(targets.size()))
+            if (targetIndex >= 0 && targetIndex < static_cast<int>(targets.size())) {
+                currentControlData().flushAndSingleExit(*this, *targets[targetIndex], m_parser->expressionStack(), false, true);
```

Comment:
How about passing `results` instead of `m_parser->expressionStack()`? They are the same.

Reply:
Looks like this is unnecessary...I thought this was the cause of some debug assertion failures, but it turns out that was a separate issue. This is resolved in the latest push.